Added infrastructure to enable easy enrichment of GenTraversables.

milessabin opened the pull request:
As discussed on [scala-internals](https://groups.google.com/d/topic/scala-internals/XRbObRJvOkU/discussion). Sample usage [here](https://gist.github.com/2500326). Motivating blog post [here](http://yz.mit.edu/wp/true-scala-complexity/).

Builds successfully and passes all tests.

milessabin committed:
Added infrastructure to enable easy enrichment of GenTraversables.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/30/

adriaanm commented:
A IsElemOf Repr?

milessabin commented:
Maybe something more literal minded would be better. Elem[A, Repr] is an alias for Repr => GenTraversableLike[A, Repr], so how about ToGenTraverasbleLike[A, Repr]? It's a bit more verbose, but OTOH it's descriptive.

adriaanm commented:
Repr HasElem A ?

milessabin commented:
That works for me. I'll update the pull request ...

milessabin commented:
Renamed Elem[A, Repr] to HasElem[Repr, A] as suggested by @adriaanm. Rebased to master, tested and squashed. Also got rid of the extraneous newline change.

retronym commented on src/library/scala/collection/generic/package.scala:
```diff
@@ -4,6 +4,12 @@ import generic.CanBuildFrom
 
+  /** The type of conversions from a collection representation type
+   *  `Repr` to its corresponding GenTraversableLike.
+   *  @see FromRepr
```
`@see [[scala.collection.generic.FromRepr]]`

milessabin replied:
Good catch ... thanks.

retronym commented on src/library/scala/collection/generic/FromRepr.scala:
```diff
@@ -0,0 +1,56 @@
+    (implicit hasElem0: HasElem[C[A0], A0]) = new FromRepr[C[A0]] {
+      type A = A0
+      val hasElem = hasElem0
+    }
```
Perhaps best to annotate the return types, e.g:

``` scala
implicit val stringFromRepr: FromRepr[String] { type A = Char} = new FromRepr[String] {
```

milessabin replied:
As a general rule I find type annotations with refinements offensive non-DRY, but given that this is an implicit definition, I have to agree.

retronym commented on src/library/scala/collection/GenTraversableLike.scala:
```diff
@@ -411,3 +411,12 @@ trait GenTraversableLike[+A, +Repr] extends Any with GenTraversableOnce[A] with
+  /** Manufacture a conversion from collection representation type `Repr` to
+   *  its corresponding `GenTraversableLike` given an implicitly available
+   *  instance of `FromRepr[Repr]`.
+   *  @see FromRepr
```
as below...

milessabin replied:
Also as below ... ;-)

adriaanm commented:
Thanks!
sorry, you'll need to submit a new pull-request for the buildbot to see it

milessabin commented:
Okey doke ... new pull request on its way. I'm closing this one.
